kola/tests/update/update.go

```diff
@@ -0,0 +1,250 @@
+
+		return fmt.Sprintf("%s:34567", srv.PrivateIP())
+	}
+}
+
+func configureMachine(m platform.Machine, c cluster.TestCluster, addr string) {
+	// update atomicly so nothing reading update.conf fails
+	c.MustSSH(m, fmt.Sprintf(`sudo bash -c "cat >/etc/coreos/update.conf.new <<EOF
+GROUP=developer
+SERVER=http://%s/v1/update
+EOF"`, addr))
+	c.MustSSH(m, "sudo mv /etc/coreos/update.conf{.new,}")
+
+	// inject dev key
+	c.MustSSH(m, `sudo bash -c "cat >/etc/coreos/update-payload-key.pub.pem <<EOF
+-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAzFS5uVJ+pgibcFLD3kbY
+k02Edj0HXq31ZT/Bva1sLp3Ysv+QTv/ezjf0gGFfASdgpz6G+zTipS9AIrQr0yFR
++tdp1ZsHLGxVwvUoXFftdapqlyj8uQcWjjbN7qJsZu0Ett/qo93hQ5nHW7Sv5dRm
+/ZsDFqk2Uvyaoef4bF9r03wYpZq7K3oALZ2smETv+A5600mj1Xg5M52QFU67UHls
+EFkZphrGjiqiCdp9AAbAvE7a5rFcJf86YR73QX08K8BX7OMzkn3DsqdnWvLB3l3W
+6kvIuP+75SrMNeYAcU8PI1+bzLcAG3VN3jA78zeKALgynUNH50mxuiiU3DO4DZ+p
+5QIDAQAB
+-----END PUBLIC KEY-----
+EOF"`)
+
+	c.MustSSH(m, "sudo mount --bind /etc/coreos/update-payload-key.pub.pem /usr/share/update_engine/update-payload-key.pub.pem")
+
+	// disable reboot so the test has explicit control
+	c.MustSSH(m, "sudo systemctl mask locksmithd.service")
+	c.MustSSH(m, "sudo systemctl stop locksmithd.service")
+	c.MustSSH(m, "sudo systemctl reset-failed locksmithd.service")
+
+	c.MustSSH(m, "sudo systemctl restart update-engine.service")
+
+	c.MustSSH(m, "update_engine --v=10")
+}
+
+func updateMachine(m platform.Machine, c cluster.TestCluster) {
+	c.Logf("Triggering update_engine")
+
+	out, stderr, err := m.SSH("update_engine_client -check_for_update")
+	if err != nil {
+		c.Fatalf("Executing update_engine_client failed: %v: %v: %s", out, err, stderr)
+	}
+
+	err = util.WaitUntilReady(120*time.Second, 10*time.Second, func() (bool, error) {
+		envs, stderr, err := m.SSH("update_engine_client -status 2>/dev/null")
+		if err != nil {
+			return false, fmt.Errorf("checking status failed: %v: %s", err, stderr)
+		}
+
+		return splitNewlineEnv(string(envs))["CURRENT_OP"] == "UPDATE_STATUS_UPDATED_NEED_REBOOT", nil
+	})
+	if err != nil {
+		c.Fatalf("Updating machine: %v", err)
+	}
+
+	c.Logf("Rebooting test machine")
+
+	if err = m.Reboot(); err != nil {
+		c.Fatalf("reboot failed: %v", err)
+	}
+}
+
+// splits newline-delimited KEY=VAL pairs into a map
+func splitNewlineEnv(envs string) map[string]string {
+	m := make(map[string]string)
+	sc := bufio.NewScanner(strings.NewReader(envs))
+	for sc.Scan() {
+		spl := strings.SplitN(sc.Text(), "=", 2)
+		m[spl[0]] = spl[1]
+	}
+	return m
+}
+
+// split space-seperated KEY=VAL pairs into a map
+func splitSpaceEnv(envs string) map[string]string {
+	m := make(map[string]string)
+	pairs := strings.Fields(envs)
+	for _, p := range pairs {
+		spl := strings.SplitN(p, "=", 2)
+		if len(spl) == 2 {
+			m[spl[0]] = spl[1]
+		}
+	}
+	return m
+}
+
+func checkUsrA(m platform.Machine, c cluster.TestCluster) {
```

Comment:
`misc` already has `assertBootedUsr()`.  Probably better to export that (maybe in a `util` package?) and reuse.

Reply:
Thoughts on location for the containing `util` package? As this requires both `cluster.TestCluster` & `platform.Machine` it cannot live in `util/` or in `platform/util.go` because of circular imports. I'm not particularly partial to creating a `util` package inside of `kola/tests` either. Maybe inside of `kola/cluster`?

Reply:
I do think it makes sense to have a `kola.tests.util`, containing purely test code shared among multiple tests.  I'm open to other ideas though.